snapcraft/plugins/jhbuild.py

```diff
@@ -0,0 +1,383 @@
+
+        self.build_packages += list(BUILD_PACKAGES)
+
+        self.jhbuild_src = os.path.join(self.partdir, 'jhbuild', 'src')
+        self.jhbuild_program = os.path.join(
+            self.partdir, 'jhbuild', 'usr', 'bin', 'jhbuild')
+        self.jhbuildrc_path = os.path.join(self.partdir, 'jhbuildrc')
+
+    def enable_cross_compilation(self):
+        """
+        We don't support cross compilation (yet?)
+        """
+        raise NotImplementedError(
+            'The {!s} plugin does not support cross compilation'.format(
+                self.name))
+
+    def run(self, cmd, cwd=None, **kwargs):
+        """Run a command.
+
+        Use run_output() if you need to capture or suppress output.
+
+        :param list cmd: command arguments, first is the executable
+        :param str cwd: working directory
+        """
+        return super().run(cmd, cwd=cwd, **kwargs)
+
+    def run_output(self, cmd, cwd=None, **kwargs):
+        """Run a command, capturing its output.
+
+        Use run() if you don't need the output and prefer it to be piped to
+        stdout.
+
+        :param list cmd: command arguments, first is the executable
+        :param str cwd: working directory
+        :return: the output of the command
+        :rtype: str
+        """
+        return super().run_output(cmd, cwd=cwd, **kwargs)
+
+    def jhbuild(self, args, output=True, **kwargs):
+        """Run JHBuild in the build stage.
+
+        :param list args: command arguments without executable
+        :return: the output of the command if captured
+        :rtype: str
+        """
+
+        cwd = kwargs.pop('cwd', os.getcwd())
+        cmd = []
+        if os.getuid() == 0:
+            cmd = ['sudo', '-H', '-u', 'jhbuild', '--']
+
+        cmd = cmd + [self.jhbuild_program,
+                     '--no-interact', '-f', self.jhbuildrc_path]
+
+        run = self.run_output if output else self.run
+        return run(cmd + args, cwd=cwd, **kwargs)
+
+    def _create_jhbuild_user(self):
+        common.run(['adduser', '--shell', '/bin/bash', '--disabled-password',
+                    '--system', '--quiet', '--home', os.sep +
+                    os.path.join(self.partdir, 'jhbuild', 'home'), 'jhbuild'])
+        return _get_jhbuild_user()
+
+    def _set_jhbuild_ownership(self, paths):
+        jhbuild_user = int(_get_jhbuild_user() or self._create_jhbuild_user())
+        jhbuild_group = int(_get_jhbuild_group())
+        for path in paths:
+            for filename in glob.iglob('%s/**' % path, recursive=True):
+                if not os.path.exists(filename):
+                    continue
+                os.chown(filename, jhbuild_user, jhbuild_group)
+
+    def _pull_jhbuild(self):
+        LOG.info('Pulling JHBuild')
+
+        repository = 'https://git.gnome.org/browse/jhbuild'
+
+        if self.options.jhbuild_mirror:
+            repository = self.options.jhbuild_mirror
+
+        if os.path.isdir(self.jhbuild_src):
```

Comment:
is the jhbuild_src dir always a git repository? And why do we need it in our own directory instead of calling it from the directory defined in the property?